Allow multiple label sets per metric

jgaskins opened the pull request:
Thanks for writing this shard. It's really well done!

The only thing that I'm stumbling over is that I can't have multiple label sets for the same metric. For example, it doesn't seem like I can emit the metrics `http_request_count{status_class="2xx"}` and `http_request_count{status_class="5xx"}` using the same registry. 

This PR adds support for that by giving each metric a data store instead of a single scalar value. Method calls on the metric delegate to the data store, which tracks all of the values for all of the label sets used.

There are some things to improve in this PR, so it's not done yet. For example, I implemented histograms and summaries in terms of counters (one for count, one for sum, and one for each bucket) because I didn't want to have to duplicate all of that logic, however that means it N+1 mutex calls (one for the histogram and one for each counter … for _each label set_), which has performance implications. I'm considering adding a `protected def Counter#inc!` method that bypasses the mutex to mitigate that.

I wrote a quick benchmark to see what the performance implications would be. As expected, it had the largest impact on histograms, which were already the slowest and are known to be heavy in Prometheus simply due to the quantity of time series they require.

<details><summary>Benchmark code</summary>

```crystal
require "benchmark"
require "../src/prometheus"

Benchmark.ips do |x|
  counter = Prometheus.counter("http_requests_total", "Total HTTP Requests")
  gauge = Prometheus.gauge "cpu_usage", "CPU usage percentage"
  histogram = Prometheus.histogram "http_response_time", "HTTP response times in seconds",
    buckets: [0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 2, 3, 5, 8, 13, 21, 30] of Float64
  summary = Prometheus.summary "http_response_errors", "HTTP Response Errors"

  x.report "counter" { counter.inc }
  x.report "gauge" { gauge.set 42 }
  x.report "histogram" { histogram.observe 0.25 }
  x.report "summary" { summary.observe 1 }
end
```

</details>

### Before

```
➜  crystal-prometheus-client git:(main) crr benchmark/metrics.cr
  counter 135.30M (  7.39ns) (± 5.64%)  0.0B/op   1.05× slower
    gauge 141.67M (  7.06ns) (± 3.89%)  0.0B/op        fastest
histogram   8.08M (123.75ns) (± 6.17%)  0.0B/op  17.53× slower
  summary 140.29M (  7.13ns) (± 3.20%)  0.0B/op   1.01× slower
```

### After

```
➜  crystal-prometheus-client git:(allow-multiple-labelsets-per-metric) crr benchmark/metrics.cr
  counter  85.96M ( 11.63ns) (± 2.59%)  0.0B/op   1.19× slower
    gauge 102.30M (  9.77ns) (± 3.82%)  0.0B/op        fastest
histogram   3.16M (316.75ns) (± 0.41%)  0.0B/op  32.40× slower
  summary  33.54M ( 29.81ns) (± 2.55%)  0.0B/op   3.05× slower
```

I don't think much can be done to improve the `Counter` and `Gauge` types with this PR. Adding 2ns of overhead per hash dereference (there's more of a difference for `Counter` than `Gauge` because it does a read _and_ a write via the data store hash whereas `Gauge` just does a write) is unavoidable. Removing the nested/repeated mutex usage should improve the histogram and summary performance dramatically, though. In other benchmarks I've run, I've found that each mutex lock is around 10ns. I added an unreasonable number of buckets to the histogram (the old rule of thumb for Prometheus was a cardinality of 10-12 and this histogram has 22 without any labels) as a pathological case to see what the performance hit would be there.

jgaskins committed:
Allow multiple label sets per metric

pfeifferj commented:
hiya! thank you for the thoughtful PR, I really appreciate it! let me know it's ready for review :)

jgaskins committed:
Mitigate nested mutex usage

jgaskins commented:
TL;DR: This is ready for review and, while it's still slower than on `main`, it's 33% faster than it was last night.

---

After https://github.com/pfeifferj/crystal-prometheus-client/pull/1/commits/56319126bce133f436d2ce5bd94e2238125909df I was able to reduce histogram and summary CPU costs by about 1/3. This is the before and after for that commit specifically:

```
➜  crystal-prometheus-client git:(allow-multiple-labelsets-per-metric) crr benchmark/metrics.cr
  counter  83.90M ( 11.92ns) (± 0.76%)  0.0B/op   1.19× slower
    gauge  99.95M ( 10.00ns) (± 1.28%)  0.0B/op        fastest
histogram   3.06M (326.73ns) (± 1.21%)  0.0B/op  32.66× slower
  summary  33.11M ( 30.20ns) (± 2.98%)  0.0B/op   3.02× slower
➜  crystal-prometheus-client git:(allow-multiple-labelsets-per-metric) crr benchmark/metrics.cr
  counter  83.87M ( 11.92ns) (± 2.57%)  0.0B/op   1.21× slower
    gauge 101.31M (  9.87ns) (± 0.99%)  0.0B/op        fastest
histogram   4.35M (229.78ns) (± 0.70%)  0.0B/op  23.28× slower
  summary  52.39M ( 19.09ns) (± 0.44%)  0.0B/op   1.93× slower
```

It's still a fair bit slower than the current `main` branch, so I ran this code through macOS Instruments.app:

```crystal
require "../src/prometheus"

histogram = Prometheus.histogram "foo", "",
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10] of Float64

loop do
  histogram.observe 0.15
end
```

What I found was that 52% of CPU time after the latest commit was spent in `Hash(String, String)#[]=`, which aligns pretty closely with my benchmarks — growth from 123ns (for the single buckets hash as implemented on `main`) to 229ns behind a second hash (as of 56319126bce133f436d2ce5bd94e2238125909df) means the gain is 46% of the total time spent, so nearly all of that gain was due specifically to a second layer of hash iterations and assignments.

jgaskins commented on src/metrics.cr:
```diff
@@ -117,52 +100,59 @@ module Prometheus
+        @count.inc! 1, labels
+        @sum.inc! value, labels
+        @buckets.each do |upper_bound, bucket|
+          if value <= upper_bound
+            inc = 1
+          else
+            inc = 0
           end
+          bucket.inc! inc, labels
```
I would love to find a better way to make the bucket increment ensure all buckets are present for all label sets. This should improve `Histogram#observe` times.

I feel like there's a way we can do it just in `collect` (because that's what this is for anyway), but my efforts there were not successful.

pfeifferj replied:
what if we moved the bucket initialization to `collect` and made `observe` more efficient? we could:

1. track observed label sets in a `@observed_labels` set
2. in `observe`, just track the label set and increment buckets where needed
3. in `collect`, ensure all buckets exist for all observed label sets by calling `get` on each bucket's store

this way we're only initializing buckets when we actually need them (during collection) rather than on every observation & the `observe` method would be faster since it doesn't need to check every bucket, and we still maintain all the same functionality

what are your thoughts on this?

jgaskins replied:
Okay, _finally_ managed to get this idea working correctly. I kept running into issues with the label sets either not propagating or propagating incorrectly.

Here's the latest benchmark as of 2a246c6:

```
              counter  96.89M ( 10.32ns) (± 1.86%)    0.0B/op    1.20× slower
                gauge 116.23M (  8.60ns) (± 0.83%)    0.0B/op         fastest
            histogram   6.19M (161.58ns) (± 0.56%)    0.0B/op   18.78× slower
              summary  64.22M ( 15.57ns) (± 1.00%)    0.0B/op    1.81× slower
```

That brings the histogram down from 230ns to 162ns. That's a lot closer to the 124ns latency that the current `main` branch has than any of my previous iterations. It also shaves off a little bit of time for counters and gauges (~1ns). I don't know how, tbh, but maybe it's due to other optimizations LLVM was able to make with this change. Regardless of the reason, since histograms and summaries are implemented in terms of counters and gauges, that compounds the performance boost. 🎉

pfeifferj replied:
awesome! shall I have one final lookover and then merge?

jgaskins replied:
Yes, it should be good to go! Thank you!

jgaskins commented on src/types.cr:
```diff
@@ -86,21 +99,46 @@ module Prometheus
 
+    protected getter store : DataStore = DataStore.new
+
+    private def label_set_for(labels : Hash(String, String)) : LabelSet
+      label_set_for LabelSet.new(labels)
+    end
+
+    private def label_set_for(labels : LabelSet) : LabelSet
+      @labels.merge(labels)
+    end
+
+    private def label_set_for(labels : Nil) : LabelSet
+      @labels
+    end
```
This is a convenience layer for consumers. It lets an application call `counter.inc 1, {"foo" => "bar"}` without having to wrap the hash in a `LabelSet`. Since that `LabelSet` would be thrown away after the merge anyway, this avoids a heap allocation and allows the application code to be more concise.

If they have no labels to add, the `nil` default avoids any overhead of merging.

pfeifferj replied:
love this!

jgaskins commented on src/types.cr:
```diff
@@ -86,21 +99,46 @@ module Prometheus
     abstract def type : String
-    abstract def collect : Array(Sample)
+
+    def collect : Array(Sample)
+      store.map do |label_set, value|
+        Sample.new(@name, label_set, value)
+      end
+    end
```
I promoted this method up to the `Metric` superclass because it made `Counter` and `Gauge` simpler. The implementations were the same. I don't know if this was the right call because `Histogram` and `Summary` have their own special implementations.

pfeifferj replied:
would it make sense if we kept the base `collect` implementation in Metric but made it `protected` instead of `def`? that way:

1. Counter and Gauge could still use the simple implementation by just calling `super`
2. Histogram and Summary could override it with their special implementations

the base implementation is still useful for simple metrics, but making it protected would make it clearer that it's meant to be used by subclasses rather than as a public interface. so we maintain the DRY principle for the simple cases while allowing flexibility for complex ones

what do you think about this approach? does it make sense?

jgaskins commented on src/types.cr:
````diff
@@ -35,21 +35,23 @@ module Prometheus
-  # ```crystal
+  # ```
   # labels = LabelSet.new({
   #   "method" => "GET",
-  #   "path" => "/api/users"
+  #   "path"   => "/api/users",
   # })
   # ```
   #
   # LabelSets can be merged to combine labels from different sources:
   #
-  # ```crystal
+  # ```
````
I didn't make these changes myself. This was the result of `crystal tool format`. I'm happy to undo them if you'd prefer they go into another PR.

pfeifferj replied:
I'm fine with using what the formatter puts out. no need to change this back

jgaskins commented on src/registry.cr:
```diff
@@ -81,26 +81,29 @@ module Prometheus
-            metric.collect.each do |sample|
-              io << sample << "\n"
-            end
+      String.build do |io|
+        collect io
+      end
+    end
+
+    def collect(io : IO) : Nil
+      @mutex.synchronize do
+        @metrics.each do |_, metric|
+          # Write help comment
+          io << "# HELP " << metric.name << " " << metric.help << "\n"
+          # Write type comment
+          io << "# TYPE " << metric.name << " " << metric.type << "\n"
+          # Write samples
+          metric.collect.each do |sample|
+            io << sample << "\n"
           end
         end
       end
-      output
     end
```
I committed this by accident. It was intended to go into another PR. This is what happens when I'm up coding at 2am. 😄

The idea here was to run `Registry#collect` and write it directly to the [`HTTP::Server::Response`](https://crystal-lang.org/api/1.16.0/HTTP/Server/Response.html) without building an intermediate string in memory. Not all frameworks allow this, but [my `armature` shard](https://github.com/jgaskins/armature) does. I could run `Prometheus.collect response` and it'll create the arrays of `Sample`s, but not the serialization.

I've also been thinking of other ways to avoid heap allocations, such as for those arrays of samples. Each array involves 2 heap allocations: 1 for the array itself since it's a `class` type and 1 for the [buffer](https://github.com/crystal-lang/crystal/blob/53cead6befd1f86c061d40e659892c9c0b84b5e3/src/array.cr#L77). Each `Sample` could probably be a `struct` type since it doesn't accumulate state. Small things like that.

pfeifferj commented:
I tested things with:
crystal `1.16.1`
ameba `1.4.3`

and they worked fine. you might want to remove the `version: ~> 1.4.0` version constraint for ameba because that caused errors for me.

looking at the code changes now

jgaskins committed:
Lazily initialize label sets for histogram buckets

Doing it eagerly in `observe` was more expensive than we needed it to
be. We only need the values there when we're emitting the metrics to
the Prometheus scraper, so we can just do it then.

jgaskins commented:
Awesome, thank you for reviewing!
